src/main/java/io/supertokens/saml/SAML.java

```diff
@@ -0,0 +1,690 @@
+            // Check NotBefore
+            if (assertion.getConditions() != null && assertion.getConditions().getNotBefore() != null) {
+                if (now.isBefore(assertion.getConditions().getNotBefore())) {
+                    throw new SAMLResponseVerificationFailedException();
+                }
+            }
+
+            // Check NotOnOrAfter
+            if (assertion.getConditions() != null && assertion.getConditions().getNotOnOrAfter() != null) {
+                if (now.isAfter(assertion.getConditions().getNotOnOrAfter())) {
+                    throw new SAMLResponseVerificationFailedException();
+                }
+            }
+        }
+    }
+
+    public static String handleCallback(Main main, TenantIdentifier tenantIdentifier, Storage storage, String samlResponse, String relayState)
+            throws StorageQueryException, XMLParserException, IOException, UnmarshallingException,
+            CertificateException, InvalidRelayStateException, SAMLResponseVerificationFailedException,
+            InvalidClientException, IDPInitiatedLoginDisallowedException, TenantOrAppNotFoundException,
+            FeatureNotEnabledException {
+        checkForSAMLFeature(tenantIdentifier.toAppIdentifier(), main);
+
+        SAMLStorage samlStorage = StorageUtils.getSAMLStorage(storage);
+        CoreConfig config = Config.getConfig(tenantIdentifier, main);
+
+        SAMLClient client = null;
+        Response response = parseSamlResponse(samlResponse);
+        String state = null;
+        String redirectURI = null;
+
+        if (relayState != null && !relayState.isEmpty()) {
+            // sp initiated
+            var relayStateInfo = samlStorage.getRelayStateInfo(tenantIdentifier, relayState);
+            if (relayStateInfo == null) {
+                throw new InvalidRelayStateException();
+            }
+
+            String clientId = relayStateInfo.clientId;
+            client = samlStorage.getSAMLClient(tenantIdentifier, clientId);
+            state = relayStateInfo.state;
+            redirectURI = relayStateInfo.redirectURI;
+        } else {
+            // idp initiated
+            String idpEntityId = response.getIssuer().getValue();
+            client = samlStorage.getSAMLClientByIDPEntityId(tenantIdentifier, idpEntityId);
+            redirectURI = client.defaultRedirectURI;
+
+            if (!client.allowIDPInitiatedLogin) {
+                throw new IDPInitiatedLoginDisallowedException();
+            }
+        }
+
+        if (client == null) {
+            throw new InvalidClientException();
+        }
+
+        // SAML verification
+        X509Certificate idpSigningCertificate = getCertificateFromString(client.idpSigningCertificate);
+        try {
+            verifySamlResponseSignature(response, idpSigningCertificate);
+        } catch (SignatureException e) {
+            throw new SAMLResponseVerificationFailedException();
+        }
+        validateSamlResponseTimestamps(response);
```

Comment:
Do we have a checklist in the standard for all the checks we need in each step that we can verify?